### ISSUE TITLE
Question about Comparative Learning

### ISSUE THREAD

**Issue by Tshning** [laichengen/MCLE#3]
Firstly, I would like to express my thoughts on your sharing. I have a bit of confusion about comparative learning:

loss_sem = self.cl_loss(img_ques_cap_proj, ans_proj)

def cl_loss(self, x, y, score=None):
        device = x.device
        sim = x @ y.T / self.temp
        labels = torch.arange(sim.shape[0], device=device, dtype=torch.long)
        return self.lossfn(sim, labels, score) 

class CELoss(nn.Module):
        def __init__(self):
              super().__init__()
        def forward(self, pred, target, hard_negative_weight=None):
              eps = 1e-12
              # standard cross entropy loss
              if hard_negative_weight is not None:
              loss = -1. * pred.gather(1, target.unsqueeze(-1)) + torch.log(
              torch.exp(pred * hard_negative_weight + eps).sum(dim=1))
              else:
              loss = -1. * pred.gather(1, target.unsqueeze(-1)) + torch.log(torch.exp(pred + eps).sum(dim=1))
              # loss = -1. * pred.gather(1, target.unsqueeze(-1)) + torch.log(torch.exp(pred + eps) * hard_negative_weight.sum(dim=1))
              return loss.mean()
              
Does this mean that, with img_ques_cap_proj as the anchor point, ans_proj extracted from text_imbed is the positive sample, while ans_proj from other samples in the same batch is used as the negative sample to calculate the contrastive loss
